drivers/gpio/gpio_rda5981a.c

```diff
@@ -0,0 +1,230 @@
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int gpio_rda5981a_manage_callback(struct device *dev,
+				    struct gpio_callback *callback, bool set)
+{
+	struct gpio_rda5981a_dev_data *data = DEV_GPIO_DATA(dev);
+
+	_gpio_manage_callback(&data->callbacks, callback, set);
+
+	return 0;
+}
+
+
+static int gpio_rda5981a_enable_callback(struct device *dev,
+				    int access_op, uint32_t pin)
+{
+	struct gpio_rda5981a_dev_data *data = DEV_GPIO_DATA(dev);
+
+	if (access_op == GPIO_ACCESS_BY_PIN) {
+		data->pin_callback_enables |= (1 << pin);
+	} else {
+		data->pin_callback_enables = 0xFFFFFFFF;
+	}
+
+	return 0;
+}
+
+static int gpio_rda5981a_disable_callback(struct device *dev,
+				     int access_op, uint32_t pin)
+{
+	struct gpio_rda5981a_dev_data *data = DEV_GPIO_DATA(dev);
+
+	if (access_op == GPIO_ACCESS_BY_PIN) {
+		data->pin_callback_enables &= ~(1 << pin);
+	} else {
+		data->pin_callback_enables = 0;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Handler for port interrupts
+ * @param dev Pointer to device structure for driver instance
+ *
+ * @return N/A
+ */
+static void gpio_rda5981a_port_isr(void *arg)
+{
+	uint32_t enabled_int, int_status = 0;
+	struct device *dev = arg;
+	struct gpio_rda5981a_dev_data *data = DEV_GPIO_DATA(dev);
+	volatile struct gpio_rda5981a *gpio = GPIO_STRUCT(dev);
+
+	int_status = (gpio->intctrl & GPIO_INT0_EN_MASK);
+	enabled_int = int_status & data->pin_callback_enables;
+
+	irq_disable(GPIO_IRQ);
+
+	/* Call the registered callbacks */
+	_gpio_fire_callbacks(&data->callbacks, (struct device *)dev,
+			     enabled_int);
+
+	irq_enable(GPIO_IRQ);
+}
+
+static const struct gpio_driver_api gpio_rda5981a_drv_api_funcs = {
+	.config = gpio_rda5981a_config,
+	.read = gpio_rda5981a_read,
+	.write = gpio_rda5981a_write,
+	.manage_callback = gpio_rda5981a_manage_callback,
+	.enable_callback = gpio_rda5981a_enable_callback,
+	.disable_callback = gpio_rda5981a_disable_callback,
+};
+
+static const struct gpio_rda5981a_cfg gpio_cfg = {
+	.gpio_base_addr   = RDA_GPIO_BASE,
+};
+
+static struct gpio_rda5981a_dev_data gpio_data;
+
+static int gpio_rda5981a_init(struct device *dev);
+
+DEVICE_AND_API_INIT(gpio_rda5981a, CONFIG_GPIO_RDA5981A_DEV_NAME,
+		    gpio_rda5981a_init,
+		    &gpio_data, &gpio_cfg,
+		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
+		    &gpio_rda5981a_drv_api_funcs);
+
+#define SPLASH_LED 1
+#if SPLASH_LED
+#define SPLASH_STACK_SIZE	256
+static char __noinit __stack splash_stack[SPLASH_STACK_SIZE];
+static void splash(void *arg1, void *arg2, void *arg3)
+{
+	int toggle = 1;
+	struct device *dev = (struct device *)arg1;
+
+	gpio_rda5981a_config(dev, GPIO_ACCESS_BY_PIN, GPIO_PIN2,
+				(GPIO_DIR_OUT));
+	gpio_rda5981a_config(dev, GPIO_ACCESS_BY_PIN, GPIO_PIN3,
+				(GPIO_DIR_OUT));
+
+	while (1) {
+		gpio_rda5981a_write(dev, GPIO_ACCESS_BY_PIN,
+					GPIO_PIN2, toggle);
+		gpio_rda5981a_write(dev, GPIO_ACCESS_BY_PIN,
+					GPIO_PIN3, !toggle);
+		k_sleep(1000);
+		toggle = toggle ? 0 : 1;
```

Comment:
Why is this splash thread in the gpio driver? It looks more like a sample or test application